File: examples/ruby/cosmwasm_v1_transaction_messages.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# cosmwasm.wasm.v1 Transaction Messages — Protobuf Examples
#
# This file demonstrates how to construct, encode, and decode
# transaction messages for CosmWasm smart contracts.
#
# Messages covered:
# - MsgExecuteContract: Execute a method on a deployed contract
# - MsgInstantiateContract: Deploy a new contract instance from stored code
# - MsgMigrateContract: Upgrade a contract to new code
# - MsgUpdateAdmin: Change contract admin
# - MsgClearAdmin: Remove contract admin (makes contract immutable)

# This example uses the xion_types gem. Install it with:
#   gem install xion_types
# Or add it to your Gemfile:
#   gem 'xion_types'

require 'xion_types'
require 'json'

# =============================================================================
# MsgExecuteContract
# =============================================================================
# Execute a method on a deployed smart contract.
# Type URL: /cosmwasm.wasm.v1.MsgExecuteContract
# Signer: sender

def msg_execute_contract_examples
  puts '=== MsgExecuteContract Examples ==='
  puts

  # Example 1: Basic contract execution (no funds attached)
  basic_execute_msg = {
    'increment' => {}
  }

  basic_msg = Cosmwasm::Wasm::V1::MsgExecuteContract.new(
    sender: 'xion1sender...',
    contract: 'xion1contractaddress...',
    msg: basic_execute_msg.to_json.encode('utf-8'),
    funds: []
  )

  puts '1. Basic MsgExecuteContract (no funds):'
  puts "   sender: #{basic_msg.sender}"
  puts "   contract: #{basic_msg.contract}"
  puts "   msg: #{JSON.parse(basic_msg.msg.force_encoding('utf-8'))}"
  puts "   funds: #{basic_msg.funds.length}"
  puts '   typeUrl: /cosmwasm.wasm.v1.MsgExecuteContract'

  # Example 2: CW20 Token Transfer
  cw20_transfer_msg = {
    'transfer' => {
      'recipient' => 'xion1recipient...',
      'amount' => '1000000'
    }
  }

  cw20_transfer = Cosmwasm::Wasm::V1::MsgExecuteContract.new(
    sender: 'xion1sender...',
    contract: 'xion1cw20tokencontract...',
    msg: cw20_transfer_msg.to_json.encode('utf-8'),
    funds: []
  )

  puts "\n2. CW20 Token Transfer:"
  puts "   contract: #{cw20_transfer.contract}"
  puts "   msg: #{JSON.parse(cw20_transfer.msg.force_encoding('utf-8'))}"

  # Example 3: Contract execution WITH attached funds
  buy_nft_msg = {
    'buy_nft' => {
      'token_id' => '42'
    }
  }

  msg_with_funds = Cosmwasm::Wasm::V1::MsgExecuteContract.new(
    sender: 'xion1buyer...',
    contract: 'xion1nftmarketplace...',
    msg: buy_nft_msg.to_json.encode('utf-8'),
    funds: [
      Cosmos::Base::V1beta1::Coin.new(denom: 'uxion', amount: '5000000')
    ]
  )

  puts "\n3. Contract execution with funds:"
  puts "   msg: #{JSON.parse(msg_with_funds.msg.force_encoding('utf-8'))}"
  puts "   funds: #{msg_with_funds.funds[0].amount} #{msg_with_funds.funds[0].denom}"

  # Encode
  encoded_bytes = Cosmwasm::Wasm::V1::MsgExecuteContract.encode(msg_with_funds)
  puts "\n4. Encoded: #{encoded_bytes.length} bytes"

  msg_with_funds
end

# =============================================================================
# MsgInstantiateContract
# =============================================================================
# Deploy a new contract instance from stored code.
# Type URL: /cosmwasm.wasm.v1.MsgInstantiateContract
# Signer: sender

def msg_instantiate_contract_examples
  puts "\n=== MsgInstantiateContract Examples ==="
  puts

  init_msg = {
    'name' => 'My Token',
    'symbol' => 'MTK',
    'decimals' => 6,
    'initial_balances' => [
      { 'address' => 'xion1holder...', 'amount' => '1000000000' }
    ]
  }

  msg = Cosmwasm::Wasm::V1::MsgInstantiateContract.new(
    sender: 'xion1deployer...',
    admin: 'xion1admin...',
    code_id: 1,
    label: 'my-token-v1',
    msg: init_msg.to_json.encode('utf-8'),
    funds: [
      Cosmos::Base::V1beta1::Coin.new(denom: 'uxion', amount: '1000000')
    ]
  )

  puts '1. Constructed MsgInstantiateContract:'
  puts "   sender: #{msg.sender}"
  puts "   admin: #{msg.admin}"
  puts "   code_id: #{msg.code_id}"
  puts "   label: #{msg.label}"
  puts "   msg: #{JSON.parse(msg.msg.force_encoding('utf-8'))}"
  puts "   funds: #{msg.funds[0].amount} #{msg.funds[0].denom}"
  puts '   typeUrl: /cosmwasm.wasm.v1.MsgInstantiateContract'

  encoded_bytes = Cosmwasm::Wasm::V1::MsgInstantiateContract.encode(msg)
  puts "\n2. Encoded: #{encoded_bytes.length} bytes"

  msg
end

# =============================================================================
# MsgMigrateContract
# =============================================================================
# Upgrade a contract to new code.
# Type URL: /cosmwasm.wasm.v1.MsgMigrateContract
# Signer: sender (must be admin)

def msg_migrate_contract_examples
  puts "\n=== MsgMigrateContract Examples ==="
  puts

  migrate_msg = {
    'migrate' => { 'new_field' => 'value' }
  }

  msg = Cosmwasm::Wasm::V1::MsgMigrateContract.new(
    sender: 'xion1admin...',
    contract: 'xion1contractaddress...',
    code_id: 2,
    msg: migrate_msg.to_json.encode('utf-8')
  )

  puts '1. Constructed MsgMigrateContract:'
  puts "   sender: #{msg.sender}"
  puts "   contract: #{msg.contract}"
  puts "   code_id: #{msg.code_id}"
  puts "   msg: #{JSON.parse(msg.msg.force_encoding('utf-8'))}"
  puts '   typeUrl: /cosmwasm.wasm.v1.MsgMigrateContract'

  encoded_bytes = Cosmwasm::Wasm::V1::MsgMigrateContract.encode(msg)
  puts "\n2. Encoded: #{encoded_bytes.length} bytes"

  msg
end

# =============================================================================
# MsgUpdateAdmin
# =============================================================================
# Change contract admin.
# Type URL: /cosmwasm.wasm.v1.MsgUpdateAdmin
# Signer: sender (must be current admin)

def msg_update_admin_examples
  puts "\n=== MsgUpdateAdmin Examples ==="
  puts

  msg = Cosmwasm::Wasm::V1::MsgUpdateAdmin.new(
    sender: 'xion1currentadmin...',
    new_admin: 'xion1newadmin...',
    contract: 'xion1contractaddress...'
  )

  puts '1. Constructed MsgUpdateAdmin:'
  puts "   sender: #{msg.sender}"
  puts "   new_admin: #{msg.new_admin}"
  puts "   contract: #{msg.contract}"
  puts '   typeUrl: /cosmwasm.wasm.v1.MsgUpdateAdmin'

  encoded_bytes = Cosmwasm::Wasm::V1::MsgUpdateAdmin.encode(msg)
  puts "\n2. Encoded: #{encoded_bytes.length} bytes"

  msg
end

# =============================================================================
# MsgClearAdmin
# =============================================================================
# Remove contract admin (makes contract immutable).
# Type URL: /cosmwasm.wasm.v1.MsgClearAdmin
# Signer: sender (must be current admin)

def msg_clear_admin_examples
  puts "\n=== MsgClearAdmin Examples ==="
  puts

  msg = Cosmwasm::Wasm::V1::MsgClearAdmin.new(
    sender: 'xion1admin...',
    contract: 'xion1contractaddress...'
  )

  puts '1. Constructed MsgClearAdmin:'
  puts "   sender: #{msg.sender}"
  puts "   contract: #{msg.contract}"
  puts '   typeUrl: /cosmwasm.wasm.v1.MsgClearAdmin'
  puts "\n   WARNING: This makes the contract immutable!"

  encoded_bytes = Cosmwasm::Wasm::V1::MsgClearAdmin.encode(msg)
  puts "\n2. Encoded: #{encoded_bytes.length} bytes"

  msg
end

# =============================================================================
# Run All Examples
# =============================================================================

def main
  puts '╔════════════════════════════════════════════════════════════╗'
  puts '║  cosmwasm.wasm.v1 Transaction Messages — Protobuf Examples ║'
  puts '╚════════════════════════════════════════════════════════════╝'
  puts

  msg_execute_contract_examples
  msg_instantiate_contract_examples
  msg_migrate_contract_examples
  msg_update_admin_examples
  msg_clear_admin_examples

  puts "\n═══════════════════════════════════════════════════════════════"
  puts 'Type URL Reference:'
  puts '  MsgExecuteContract:      /cosmwasm.wasm.v1.MsgExecuteContract'
  puts '  MsgInstantiateContract:  /cosmwasm.wasm.v1.MsgInstantiateContract'
  puts '  MsgMigrateContract:      /cosmwasm.wasm.v1.MsgMigrateContract'
  puts '  MsgUpdateAdmin:          /cosmwasm.wasm.v1.MsgUpdateAdmin'
  puts '  MsgClearAdmin:           /cosmwasm.wasm.v1.MsgClearAdmin'
  puts '═══════════════════════════════════════════════════════════════'
end

# Execute if run directly
main if __FILE__ == $PROGRAM_NAME

File: examples/python/cosmwasm_v1_transaction_messages.py
#!/usr/bin/env python3
"""
cosmwasm.wasm.v1 Transaction Messages — Protobuf Examples

This file demonstrates how to construct, encode, and decode
transaction messages from the CosmWasm Module (cosmwasm.wasm.v1).

Messages covered:
- MsgStoreCode: Upload a WASM contract
- MsgInstantiateContract: Create a new contract instance
- MsgInstantiateContract2: Create a contract with predictable address
- MsgExecuteContract: Execute a contract function
- MsgMigrateContract: Migrate a contract to new code
- MsgUpdateAdmin: Update contract admin
- MsgClearAdmin: Remove contract admin
"""

import sys
import os
import json

# Add the python types to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python/types'))

from cosmwasm.wasm.v1 import tx_pb2 as wasm_tx
from google.protobuf.json_format import MessageToJson, MessageToDict


# =============================================================================
# MsgStoreCode
# =============================================================================
# Upload a new WASM contract to the blockchain.
# Type URL: /cosmwasm.wasm.v1.MsgStoreCode
# Signer: sender

def msg_store_code_examples():
    print('=== MsgStoreCode Examples ===\n')

    # 1. Construct message
    msg = wasm_tx.MsgStoreCode()
    msg.sender = 'xion1deployer...'
    # In practice, wasm_byte_code would be the actual compiled contract bytes
    msg.wasm_byte_code = b'\x00\x61\x73\x6d\x01\x00\x00\x00'  # Minimal WASM header
    # Optional: Restrict who can instantiate this code
    # msg.instantiate_permission.permission = 1  # AccessTypeOnlyAddress
    # msg.instantiate_permission.address = 'xion1allowed...'

    print('1. Constructed MsgStoreCode:')
    print(f'   sender: {msg.sender}')
    print(f'   wasm_byte_code length: {len(msg.wasm_byte_code)} bytes')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgStoreCode')

    # 2. Encode to protobuf bytes
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    # 3. Decode
    decoded = wasm_tx.MsgStoreCode()
    decoded.ParseFromString(encoded_bytes)
    print(f'\n3. Decoded sender: {decoded.sender}')
    print(f'   Decoded wasm_byte_code length: {len(decoded.wasm_byte_code)} bytes')

    return msg


# =============================================================================
# MsgInstantiateContract
# =============================================================================
# Create a new instance of an uploaded contract.
# Type URL: /cosmwasm.wasm.v1.MsgInstantiateContract
# Signer: sender

def msg_instantiate_contract_examples():
    print('\n=== MsgInstantiateContract Examples ===\n')

    # 1. Construct instantiate message
    init_msg = json.dumps({
        'name': 'My Token',
        'symbol': 'MTK',
        'decimals': 6,
        'initial_balances': [
            {'address': 'xion1holder...', 'amount': '1000000000'}
        ]
    }).encode('utf-8')

    # 2. Construct MsgInstantiateContract
    msg = wasm_tx.MsgInstantiateContract()
    msg.sender = 'xion1deployer...'
    msg.admin = 'xion1admin...'  # Optional admin for migrations
    msg.code_id = 1  # Code ID from MsgStoreCode
    msg.label = 'my-token-v1'
    msg.msg = init_msg
    # Optional: Send funds to contract on instantiation
    fund = msg.funds.add()
    fund.denom = 'uxion'
    fund.amount = '1000000'

    print('1. Constructed MsgInstantiateContract:')
    print(f'   sender: {msg.sender}')
    print(f'   admin: {msg.admin}')
    print(f'   code_id: {msg.code_id}')
    print(f'   label: {msg.label}')
    print(f'   msg: {msg.msg.decode("utf-8")}')
    print(f'   funds: {msg.funds[0].amount} {msg.funds[0].denom}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgInstantiateContract')

    # 3. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    # 4. Decode
    decoded = wasm_tx.MsgInstantiateContract()
    decoded.ParseFromString(encoded_bytes)
    print(f'\n3. Decoded label: {decoded.label}')
    print(f'   Decoded code_id: {decoded.code_id}')

    return msg


# =============================================================================
# MsgInstantiateContract2
# =============================================================================
# Create a contract with a predictable address (using salt).
# Type URL: /cosmwasm.wasm.v1.MsgInstantiateContract2
# Signer: sender

def msg_instantiate_contract2_examples():
    print('\n=== MsgInstantiateContract2 Examples ===\n')

    init_msg = json.dumps({'count': 0}).encode('utf-8')

    msg = wasm_tx.MsgInstantiateContract2()
    msg.sender = 'xion1deployer...'
    msg.admin = 'xion1admin...'
    msg.code_id = 1
    msg.label = 'counter-v1'
    msg.msg = init_msg
    msg.salt = b'unique-salt-12345'  # Makes address deterministic
    msg.fix_msg = False  # Whether to include msg in address derivation

    print('1. Constructed MsgInstantiateContract2:')
    print(f'   sender: {msg.sender}')
    print(f'   code_id: {msg.code_id}')
    print(f'   label: {msg.label}')
    print(f'   salt: {msg.salt}')
    print(f'   fix_msg: {msg.fix_msg}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgInstantiateContract2')

    # 2. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    return msg


# =============================================================================
# MsgExecuteContract
# =============================================================================
# Execute a function on an existing contract.
# Type URL: /cosmwasm.wasm.v1.MsgExecuteContract
# Signer: sender

def msg_execute_contract_examples():
    print('\n=== MsgExecuteContract Examples ===\n')

    # 1. Construct execute message (contract-specific JSON)
    execute_msg = json.dumps({
        'transfer': {
            'recipient': 'xion1recipient...',
            'amount': '1000000'
        }
    }).encode('utf-8')

    # 2. Construct MsgExecuteContract
    msg = wasm_tx.MsgExecuteContract()
    msg.sender = 'xion1sender...'
    msg.contract = 'xion1contractaddress...'
    msg.msg = execute_msg
    # Optional: Send funds with execution
    fund = msg.funds.add()
    fund.denom = 'uxion'
    fund.amount = '500000'

    print('1. Constructed MsgExecuteContract:')
    print(f'   sender: {msg.sender}')
    print(f'   contract: {msg.contract}')
    print(f'   msg: {msg.msg.decode("utf-8")}')
    print(f'   funds: {msg.funds[0].amount} {msg.funds[0].denom}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgExecuteContract')

    # 3. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')
    print(f'   Hex: {encoded_bytes.hex()[:80]}...')

    # 4. Decode
    decoded = wasm_tx.MsgExecuteContract()
    decoded.ParseFromString(encoded_bytes)
    print(f'\n3. Decoded contract: {decoded.contract}')
    print(f'   Decoded msg: {decoded.msg.decode("utf-8")}')

    # 5. JSON format
    json_str = MessageToJson(msg)
    print(f'\n4. JSON format (truncated):')
    print(f'   {json_str[:200]}...')

    return msg


# =============================================================================
# MsgMigrateContract
# =============================================================================
# Migrate a contract to a new code version.
# Type URL: /cosmwasm.wasm.v1.MsgMigrateContract
# Signer: sender (must be admin)

def msg_migrate_contract_examples():
    print('\n=== MsgMigrateContract Examples ===\n')

    # Migrate message (contract-specific)
    migrate_msg = json.dumps({
        'migrate': {'new_field': 'value'}
    }).encode('utf-8')

    msg = wasm_tx.MsgMigrateContract()
    msg.sender = 'xion1admin...'  # Must be contract admin
    msg.contract = 'xion1contractaddress...'
    msg.code_id = 2  # New code ID
    msg.msg = migrate_msg

    print('1. Constructed MsgMigrateContract:')
    print(f'   sender: {msg.sender}')
    print(f'   contract: {msg.contract}')
    print(f'   code_id: {msg.code_id}')
    print(f'   msg: {msg.msg.decode("utf-8")}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgMigrateContract')

    # 2. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    return msg


# =============================================================================
# MsgUpdateAdmin
# =============================================================================
# Update the admin of a contract.
# Type URL: /cosmwasm.wasm.v1.MsgUpdateAdmin
# Signer: sender (must be current admin)

def msg_update_admin_examples():
    print('\n=== MsgUpdateAdmin Examples ===\n')

    msg = wasm_tx.MsgUpdateAdmin()
    msg.sender = 'xion1currentadmin...'  # Current admin
    msg.new_admin = 'xion1newadmin...'
    msg.contract = 'xion1contractaddress...'

    print('1. Constructed MsgUpdateAdmin:')
    print(f'   sender: {msg.sender}')
    print(f'   new_admin: {msg.new_admin}')
    print(f'   contract: {msg.contract}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgUpdateAdmin')

    # 2. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    return msg


# =============================================================================
# MsgClearAdmin
# =============================================================================
# Remove the admin from a contract (makes it immutable).
# Type URL: /cosmwasm.wasm.v1.MsgClearAdmin
# Signer: sender (must be current admin)

def msg_clear_admin_examples():
    print('\n=== MsgClearAdmin Examples ===\n')

    msg = wasm_tx.MsgClearAdmin()
    msg.sender = 'xion1admin...'  # Current admin
    msg.contract = 'xion1contractaddress...'

    print('1. Constructed MsgClearAdmin:')
    print(f'   sender: {msg.sender}')
    print(f'   contract: {msg.contract}')
    print(f'   typeUrl: /cosmwasm.wasm.v1.MsgClearAdmin')
    print('\n   WARNING: This makes the contract immutable!')

    # 2. Encode
    encoded_bytes = msg.SerializeToString()
    print(f'\n2. Encoded: {len(encoded_bytes)} bytes')

    return msg


# =============================================================================
# Run All Examples
# =============================================================================

def main():
    print('╔════════════════════════════════════════════════════════════╗')
    print('║  cosmwasm.wasm.v1 Transaction Messages — Protobuf Examples ║')
    print('╚════════════════════════════════════════════════════════════╝\n')

    msg_store_code_examples()
    msg_instantiate_contract_examples()
    msg_instantiate_contract2_examples()
    msg_execute_contract_examples()
    msg_migrate_contract_examples()
    msg_update_admin_examples()
    msg_clear_admin_examples()

    print('\n═══════════════════════════════════════════════════════════════')
    print('Type URL Reference:')
    print('  MsgStoreCode:            /cosmwasm.wasm.v1.MsgStoreCode')
    print('  MsgInstantiateContract:  /cosmwasm.wasm.v1.MsgInstantiateContract')
    print('  MsgInstantiateContract2: /cosmwasm.wasm.v1.MsgInstantiateContract2')
    print('  MsgExecuteContract:      /cosmwasm.wasm.v1.MsgExecuteContract')
    print('  MsgMigrateContract:      /cosmwasm.wasm.v1.MsgMigrateContract')
    print('  MsgUpdateAdmin:          /cosmwasm.wasm.v1.MsgUpdateAdmin')
    print('  MsgClearAdmin:           /cosmwasm.wasm.v1.MsgClearAdmin')
    print('═══════════════════════════════════════════════════════════════')


if __name__ == '__main__':
    main()
